feat: Support custom BigQuery storage api endpoint

nj1973 opened the pull request:
## Description of changes
The PR adds support for a custom googleapis endpoint for the BigQuery Storage API and Secret Manager, in addition to the BigQuery and Spanner APIs that we already support.

I have manually tested this in a project with a PSC and firewall rules to block access to standard `googleapis.com` endpoints and the changes work for both the BigQuery Storage API and Secret Manager API. I don't think we can realistically add integration tests for this setup though.

I suspect we have Cloud Storage issues too but that is not so simple to resolve because we support using Cloud Storage for config files in addition to standard connections. I think Cloud Storage will merit its own issue.

## Issues to be closed
Closes #1369

## Checklist

- [x] I have followed the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/professional-services-data-validator/blob/develop/CONTRIBUTING.md).
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated any relevant documentation to reflect my changes, if applicable
- [x] I have added unit and/or integration tests relevant to my change as needed
- [x] I have already checked locally that all unit tests and linting are passing (use the `tests/local_check.sh` script)
- [x] I have manually executed end-to-end testing (E2E) with the affected databases/engines

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
Merge remote-tracking branch 'origin/develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
feat: Experiment with definition of custom BigQuery Storage API client

nj1973 committed:
chore: Reformat

nj1973 committed:
chore: Reformat

helensilva14 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

helensilva14 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

helensilva14 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

helensilva14 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

helensilva14 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 committed:
Merge remote-tracking branch 'origin/develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 committed:
feat: Support custom BigQuery storage api endpoint

nj1973 committed:
feat: Support custom BigQuery storage api endpoint

nj1973 committed:
feat: Support custom BigQuery storage api endpoint

nj1973 commented:
/gcbrun

nj1973 committed:
chore: reformat

nj1973 commented:
/gcbrun

nj1973 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 commented:
/gcbrun

sundar-mudupalli-work committed:
Updated and wrote a small section on running DVT at on-prem

sundar-mudupalli-work commented:
Neil,

I don't know how this is supposed to work. I am a bit confused.

Sundar Mudupalli

sundar-mudupalli-work commented on third_party/ibis/ibis_bigquery/__init__.py:
```diff
@@ -0,0 +1,137 @@
+                client_info=_create_client_info(application_name),
+            )
+        else:
+            self.client = bigquery_client
```
Hi,

I don't understand how this is going to work. According to the changes in the documentation, the user passes a string as the endpoint. Here, self.client (a python object) representing the connection to BQ is being assigned a string value. That does not sound right.

If an API endpoint is being used, based on [BigQuery client code](https://github.com/googleapis/python-bigquery/blob/a69d6b796d2edb6ba453980c9553bc9b206c5a6e/google/cloud/bigquery/client.py#L221), you need to pass the `client_options` kwarg an instance of the `ClientOptions` class while creating a connection. ClientOptions is defined in [API Core Client Options](https://googleapis.dev/python/google-api-core/latest/client_options.html)

The latest version of Ibis performs an assignment much like what is being proposed above. However, the [documentation](https://github.com/ibis-project/ibis/blob/d7cd846627269b1f4c901cc3f0e380ee9024f796/ibis/backends/bigquery/__init__.py#L394), says it is a Client (object) from `google.cloud.bigquery` package.

Based on the above analysis - one way to accomplish this might be to have DVT environment variables for API endpoints for BQ, GCS, Secret Manager and Spanner. If those are set, then we use those values when connecting to the client API's. I have started a section in the doc on running DVT at on-prem where this explanation may fit. This looks like a more complicated change. I would rather be wrong and we have a simple fix - not sure that is possible here.

I am not aware that we use BigQuery storage API. That is only for Streaming reads. 

We might be able to test this by running DVT from a laptop with an IAP tunnel to a PSC endpoint for `googleapis.com` in `pso-kokoro-resources` project. It may be too complicated to make the test part of the integration suite.

Thanks.

Sundar Mudupalli

nj1973 replied:
Regarding:
> "I am not aware that we use BigQuery storage API. That is only for Streaming reads."

I discuss this in the issue here:

https://github.com/GoogleCloudPlatform/professional-services-data-validator/issues/1369#issuecomment-2539228097

We do not directly use the storage API but Ibis does.

nj1973 replied:
Regarding:

> Here, self.client (a python object) representing the connection to BQ is being assigned a string value. That does not sound right.

The argument is a Python object, not a string. I've added type hints to make that clearer.

nj1973 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 committed:
chore: Add type hints

nj1973 commented:
/gcbrun

helensilva14 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 commented:
After discussion the new plan is:

- Test whether we can completely avoid having a storage client by passing parameters requesting no client
- If above is possible, test performance impact
- Consider Secret Manager and maybe Cloud Storage impact

nj1973 commented:
@sundar-mudupalli-work I completed some of the testing mentioned above.

**Test whether we can completely avoid having a storage client by passing parameters requesting no client**
Yes we can, with:
```
$ git diff
diff --git a/third_party/ibis/ibis_bigquery/__init__.py b/third_party/ibis/ibis_bigquery/__init__.py
index 7627398..873df3f 100644
--- a/third_party/ibis/ibis_bigquery/__init__.py
+++ b/third_party/ibis/ibis_bigquery/__init__.py
@@ -117,8 +117,10 @@ class Backend(BigQueryBackend):
                 return result.to_arrow(
                     progress_bar_type=None,
                     # Include DVT specific storage client.
-                    bqstorage_client=storage_client,
-                    create_bqstorage_client=bool(not self.storage_client),
+                    # bqstorage_client=storage_client,
+                    # create_bqstorage_client=bool(not self.storage_client),
+                    bqstorage_client=None,
+                    create_bqstorage_client=False,
                 )
```

**If above is possible, test performance impact**
Tested a 1 million row validation, Oracle to BigQuery, with the storage API code included:
```
$ time data-validation validate row -sc=oravol -tc=bq -tbls=dvt_test.tab_vol_1m -pk=id --hash='*' --filter-status=fail
╒═══════════════════╤═══════════════════╤═════════════════════╤══════════════════════╤════════════════════╤════════════════════╤══════════════════╤═════════════════════╤══════════╕
│ validation_name   │ validation_type   │ source_table_name   │ source_column_name   │ source_agg_value   │ target_agg_value   │ pct_difference   │ validation_status   │ run_id   │
╞═══════════════════╪═══════════════════╪═════════════════════╪══════════════════════╪════════════════════╪════════════════════╪══════════════════╪═════════════════════╪══════════╡
╘═══════════════════╧═══════════════════╧═════════════════════╧══════════════════════╧════════════════════╧════════════════════╧══════════════════╧═════════════════════╧══════════╛

real	0m26.995s
user	0m21.374s
sys	0m2.444s
```

Tested the same but with the storage API bypassed:
```
$ time data-validation validate row -sc=oravol -tc=bq -tbls=dvt_test.tab_vol_1m -pk=id --hash='*' --filter-status=fail
╒═══════════════════╤═══════════════════╤═════════════════════╤══════════════════════╤════════════════════╤════════════════════╤══════════════════╤═════════════════════╤══════════╕
│ validation_name   │ validation_type   │ source_table_name   │ source_column_name   │ source_agg_value   │ target_agg_value   │ pct_difference   │ validation_status   │ run_id   │
╞═══════════════════╪═══════════════════╪═════════════════════╪══════════════════════╪════════════════════╪════════════════════╪══════════════════╪═════════════════════╪══════════╡
╘═══════════════════╧═══════════════════╧═════════════════════╧══════════════════════╧════════════════════╧════════════════════╧══════════════════╧═════════════════════╧══════════╛

real	3m0.982s
user	0m28.556s
sys	0m2.724s
```

That's 3 minutes over 27 seconds. Plus I believe BigQuery Storage API reads are much cheaper than standard BigQuery queries.

I think we should retain the new code.

nj1973 committed:
feat: Add --secret-manager-api-endpoint option

nj1973 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 commented:
/gcbrun

nj1973 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 commented:
/gcbrun

sundar-mudupalli-work commented:
Neil,

Here are the results of my testing. We can discuss more on our call on
Monday.

I have a table with 24M rows - which I partitioned into 8 to have 3M rows
per partition. When I run it, I see a very small difference in performance.
You can see my connection string at the end of this message. I have given
you editor access to my project - if you want to test this yourself.

One possible explanation (not highly plausible) is that since I am using
partitioning, it is not using the storage API for read. That is a bit far
fetched for me.

Thanks

Sundar Mudupalli

(env) ***@***.***:~/issue1369$ export ENGINE=my_bq
(env) ***@***.***:~/issue1369$ data-validation
generate-table-partitions -sc $ENGINE -tc $ENGINE -tbls=dvt_test.iowa_small
-pk=invoice_and_item_number --hash='*' -cdir ~/issue1369/$ENGINE  -pn 8
 --filter-status=fail
05/04/2025 01:42:53 AM-INFO: Writing table partition configs to directory:
/home/user/issue1369/my_bq
05/04/2025 01:42:53 AM-INFO: Success! Table partition configs written to
directory: /home/user/issue1369/my_bq
(env) ***@***.***:~/issue1369$ time data-validation configs
run -c /home/user/issue1369/$ENGINE/dvt_test.iowa_small//0000.yaml
05/04/2025 01:44:36 AM-INFO: {"run_id":
"9e0eba2d-ee21-4c1f-b088-dc223f9f1478", "start_time":
"2025-05-04T01:43:11.791208+00:00", "end_time":
"2025-05-04T01:43:50.174465+00:00", "total_source_rows": 3028678,
"total_target_rows": 3028678, "total_rows_validated": 3028678,
"total_rows_success_validation_status": 3028678,
"total_rows_fail_validation_status": 0,
"failed_rows_present_in_source_not_in_target": 0,
"failed_rows_present_in_target_not_in_source": 0,
"failed_rows_present_in_both_source_and_target": 0}
╒═══════════════════╤═══════════════════╤═════════════════════╤══════════════════════╤════════════════════╤════════════════════╤══════════════════╤═════════════════════╤══════════╕
│ validation_name   │ validation_type   │ source_table_name   │
source_column_name   │ source_agg_value   │ target_agg_value   │
pct_difference   │ validation_status   │ run_id   │
╞═══════════════════╪═══════════════════╪═════════════════════╪══════════════════════╪════════════════════╪════════════════════╪══════════════════╪═════════════════════╪══════════╡
╘═══════════════════╧═══════════════════╧═════════════════════╧══════════════════════╧════════════════════╧════════════════════╧══════════════════╧═════════════════════╧══════════╛

real    1m36.604s
user    1m14.507s
sys     0m14.553s
(env) ***@***.***:~/issue1369$ export ENGINE=my_bq_stor
(env) ***@***.***:~/issue1369$ data-validation
generate-table-partitions -sc $ENGINE -tc $ENGINE -tbls=dvt_test.iowa_small
-pk=invoice_and_item_number --hash='*' -cdir ~/issue1369/$ENGINE  -pn 8
 --filter-status=fail
05/04/2025 01:45:37 AM-INFO: Writing table partition configs to directory:
/home/user/issue1369/my_bq_stor
05/04/2025 01:45:37 AM-INFO: Success! Table partition configs written to
directory: /home/user/issue1369/my_bq_stor
(env) ***@***.***:~/issue1369$ time data-validation configs
run -c /home/user/issue1369/$ENGINE/dvt_test.iowa_small//0000.yaml
05/04/2025 01:47:21 AM-INFO: {"run_id":
"fe569463-5d6f-4820-a053-e493a3304aae", "start_time":
"2025-05-04T01:46:04.142675+00:00", "end_time":
"2025-05-04T01:46:35.203262+00:00", "total_source_rows": 3028678,
"total_target_rows": 3028678, "total_rows_validated": 3028678,
"total_rows_success_validation_status": 3028678,
"total_rows_fail_validation_status": 0,
"failed_rows_present_in_source_not_in_target": 0,
"failed_rows_present_in_target_not_in_source": 0,
"failed_rows_present_in_both_source_and_target": 0}
╒═══════════════════╤═══════════════════╤═════════════════════╤══════════════════════╤════════════════════╤════════════════════╤══════════════════╤═════════════════════╤══════════╕
│ validation_name   │ validation_type   │ source_table_name   │
source_column_name   │ source_agg_value   │ target_agg_value   │
pct_difference   │ validation_status   │ run_id   │
╞═══════════════════╪═══════════════════╪═════════════════════╪══════════════════════╪════════════════════╪════════════════════╪══════════════════╪═════════════════════╪══════════╡
╘═══════════════════╧═══════════════════╧═════════════════════╧══════════════════════╧════════════════════╧════════════════════╧══════════════════╧═════════════════════╧══════════╛

real    1m32.891s
user    1m18.213s
sys     0m15.275s
(env) ***@***.***:~/issue1369$ cat
~/.config/google-pso-data-validator/my_bq_stor.connection.json
{"source_type": "BigQuery", "project_id": "mudu-plex-pri",
"storage_api_endpoint": "bigquerystorage.googleapis.com"}
(env) ***@***.***:~/issue1369$ cat
~/.config/google-pso-data-validator/my_bq.connection.json
{"source_type": "BigQuery", "project_id": "mudu-plex-pri"}

Sundar Mudupalli

On Sat, May 3, 2025 at 12:15 AM Neil Johnson ***@***.***>
wrote:

> *nj1973* left a comment
> (GoogleCloudPlatform/professional-services-data-validator#1501)
> <https://github.com/GoogleCloudPlatform/professional-services-data-validator/pull/1501#issuecomment-2848479606>
>
> /gcbrun
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/professional-services-data-validator/pull/1501#issuecomment-2848479606>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZK36WPCOYKPKM7LWCB3A2324RUIPAVCNFSM6AAAAAB26PYMM6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNBYGQ3TSNRQGY>
> .
> You are receiving this because you were mentioned.Message ID:
> <GoogleCloudPlatform/professional-services-data-validator/pull/1501/c2848479606
> @github.com>
>


-- 



Sundar Mudupalli

Enterprise Architect

Google Cloud Consulting - delta Practice

***@***.***


Upcoming OOO Jan 16-Jan 26

sundar-mudupalli-work committed:
Update connections.md

Removed references to ENDPOINT_URI - since it is not a URI, but a IP Address based on my testing. I left the references to Secret Manager in - due to my discussion in the isssue.

sundar-mudupalli-work committed:
Update README.md

Updated on-prem section to explain using endpoints from on prem.

sundar-mudupalli-work commented on data_validation/cli_tools.py:
the endpoint is a DNS name or an IP - not a URI. The comments should be updated.

nj1973 replied:
I think we need to revert this. I tried to be precise with my wording.

The Storage API requires an endpoint name but the BigQuery and Spanner clients require a URI.

For example, if I create a BigQuery client without `https://` then the call fails:
```
*** requests.exceptions.MissingSchema: Invalid URL 'bigquery-dvtpsc.p.googleapis.com/bigquery/v2/projects/dvt/datasets/pso_data_validator/tables/dvt_core_types?prettyPrint=false': No scheme supplied. Perhaps you meant https://bigquery-dvtpsc.p.googleapis.com/bigquery/v2/projects/dvt/datasets/pso_data_validator/tables/dvt_core_types?prettyPrint=false?
```

But if I include `https://` for the BigQuery Storage API then it hangs because it is not a resolvable name. I need to provide the DNS name.

I used examples that match the behaviour of the open source clients we use. It's a shame they are inconsistent.

sundar-mudupalli-work reviewed (changes requested):
Neil,

Overall fantastic work. I agree with the changes you have made, except for the secret manager API endpoint - let us discuss that.

I did not see a performance improvement from using the Storage API - I have sent you an email with access to my project on that topic. Even if there is no improvement, I think the change is good - especially if customers are using this.

On the secret manager front, I think the approach to the secret manager should be updated. I don't want to put more stuff in Connection configurations because storing them on the file system is not good, especially when they are run as containers.

Thanks.

Sundar Muduplli

nj1973 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

nj1973 committed:
docs: Docs updates

helensilva14 commented:
/gcbrun

sundar-mudupalli-work reviewed (approved):
Neil,

Excellent work!!
Apologies that you had to backout the changes I made to the doc.

Thank you.

Sundar Mudupalli

helensilva14 committed:
Merge branch 'develop' into 1369-custom-bigquery-storageapi-endpoint

helensilva14 commented:
/gcbrun

nj1973 committed:
Revert "feat: Add --secret-manager-api-endpoint option"

This reverts commit 7e3d8fb29f1b2591e5b441db5119f988064c8666.

nj1973 commented:
/gcbrun
